File: platform/packages_py/fetch_httpx/_cache/core/cache_manager.py
"""
Cache Manager

Core cache logic for managing cached responses.
"""

from __future__ import annotations

import json
import logging
import time
from typing import TYPE_CHECKING, Any

from ..._exceptions import CacheSerializationError
from ..storage.memory import MemoryStorage
from ..types import (
    CacheConfig,
    CacheEntry,
    CacheEntryMetadata,
    CacheKeyStrategy,
    CacheStats,
    RequestCacheOptions,
)
from .key_strategy import default_key_strategy

if TYPE_CHECKING:
    from ..types import CacheStorage

logger = logging.getLogger("fetch_httpx.cache")


class CacheManager:
    """
    Core cache logic for managing cached responses.

    Handles cache key generation, storage operations, and response serialization.

    Example:
        manager = CacheManager(CacheConfig(ttl=60.0))

        key = manager.generate_key('GET', 'https://api.example.com/users')
        await manager.set(key, response)

        cached = await manager.get(key)
        response = manager.create_response_from_cache(cached)
    """

    def __init__(self, config: CacheConfig | None = None) -> None:
        config = config or CacheConfig()

        self._storage: CacheStorage = config.storage or MemoryStorage(max_entries=config.max_entries)
        self._ttl = config.ttl
        self._methods = [m.upper() for m in config.methods]
        self._key_strategy: CacheKeyStrategy = config.key_strategy or default_key_strategy
        self._stale_while_revalidate = config.stale_while_revalidate
        self._stale_grace_period = config.stale_grace_period

        # Track in-flight requests for deduplication
        self._pending: dict[str, Any] = {}

    def generate_key(
        self,
        method: str,
        url: str,
        headers: dict[str, str] | None = None,
        body: Any = None,
        params: dict[str, Any] | None = None,
    ) -> str:
        """Generate a cache key for a request."""
        return self._key_strategy(method, url, headers, body, params)

    def should_cache(
        self,
        method: str,
        options: RequestCacheOptions | None = None,
    ) -> bool:
        """Check if a request should be cached."""
        if options and options.no_cache:
            return False
        return method.upper() in self._methods

    async def get(self, key: str) -> CacheEntry | None:
        """Get an entry from cache."""
        return await self._storage.get(key)

    async def get_stale(
        self, key: str
    ) -> tuple[CacheEntry | None, bool]:
        """
        Get with stale-while-revalidate support.

        Returns:
            Tuple of (entry, is_stale)
        """
        entry = await self._storage.get(key)

        if not entry:
            return None, False

        now = time.time()
        is_stale = now > entry.expires_at
        within_grace = now < entry.expires_at + self._stale_grace_period

        if is_stale and not within_grace:
            return None, True

        return entry, is_stale

    async def set(
        self,
        key: str,
        response: Any,  # Response type
        ttl: float | None = None,
    ) -> None:
        """
        Store a response in cache.

        Args:
            key: Cache key
            response: Response object to cache
            ttl: Optional TTL override

        Raises:
            CacheSerializationError: If response data cannot be serialized
        """
        effective_ttl = ttl if ttl is not None else self._ttl
        now = time.time()

        logger.info(
            "CacheManager.set called",
            extra={
                "key": key[:50],
                "ttl": effective_ttl,
                "response_type": type(response).__name__,
                "storage_type": type(self._storage).__name__,
            },
        )

        # Read response body with proper error handling
        data = None
        if hasattr(response, "json"):
            try:
                data = response.json()
            except json.JSONDecodeError as json_err:
                logger.debug(
                    "Response is not JSON, falling back to text",
                    extra={"key": key[:50], "error": str(json_err)},
                )
                # Not JSON - try text
                if hasattr(response, "text"):
                    try:
                        data = response.text
                    except Exception as text_err:
                        logger.warning(
                            "Failed to read response text",
                            extra={"key": key[:50], "error": str(text_err)},
                        )
                        data = None
            except Exception as e:
                # Unexpected error reading JSON - log and raise
                logger.error(
                    "Failed to read response body for caching",
                    extra={
                        "key": key[:50],
                        "error": str(e),
                        "error_type": type(e).__name__,
                    },
                )
                raise CacheSerializationError(
                    f"Failed to read response body: {e}",
                    operation="set",
                    key=key,
                    original_error=e,
                ) from e

        # Get metadata
        status_code = getattr(response, "status_code", 200)
        headers_obj = getattr(response, "headers", {})
        if hasattr(headers_obj, "items"):
            headers_dict = dict(headers_obj.items())
        else:
            headers_dict = dict(headers_obj) if headers_obj else {}

        url_obj = getattr(response, "url", None)
        url_str = str(url_obj) if url_obj else ""

        request_obj = getattr(response, "request", None)
        method = getattr(request_obj, "method", "GET") if request_obj else "GET"

        entry = CacheEntry(
            key=key,
            data=data,
            created_at=now,
            expires_at=now + effective_ttl,
            metadata=CacheEntryMetadata(
                status_code=status_code,
                headers=headers_dict,
                url=url_str,
                method=method,
                content_type=headers_dict.get("content-type"),
                size=len(str(data)) if data else None,
            ),
        )

        await self._storage.set(key, entry)
        logger.debug(
            "Cached response",
            extra={
                "key": key[:50],
                "status_code": status_code,
                "ttl": effective_ttl,
                "size": entry.metadata.size,
            },
        )

    def create_response_from_cache(self, entry: CacheEntry) -> Any:
        """
        Create a Response from a cached entry.

        Note: Returns a dict-like object that mimics Response.
        For full Response reconstruction, use the actual Response class.

        Raises:
            CacheSerializationError: If entry data cannot be converted to response content
        """
        # Import here to avoid circular imports
        try:
            from ..._models import URL, Headers, Response

            content: bytes
            try:
                if isinstance(entry.data, str):
                    content = entry.data.encode("utf-8")
                elif isinstance(entry.data, bytes):
                    content = entry.data
                elif entry.data is None:
                    content = b""
                else:
                    content = json.dumps(entry.data).encode("utf-8")
            except (TypeError, ValueError) as e:
                logger.error(
                    "Failed to serialize cached data to response content",
                    extra={
                        "key": entry.key[:50],
                        "data_type": type(entry.data).__name__,
                        "error": str(e),
                    },
                )
                raise CacheSerializationError(
                    f"Failed to serialize cached data: {e}",
                    operation="create_response",
                    key=entry.key,
                    original_error=e,
                ) from e

            logger.debug(
                "Created response from cache",
                extra={
                    "key": entry.key[:50],
                    "status_code": entry.metadata.status_code,
                },
            )
            return Response(
                status_code=entry.metadata.status_code,
                headers=Headers(entry.metadata.headers),
                content=content,
                request=None,
            )
        except ImportError:
            # Fallback: return a simple object
            logger.debug(
                "Using fallback CachedResponse (Response model not available)",
                extra={"key": entry.key[:50]},
            )

            class CachedResponse:
                def __init__(self, entry: CacheEntry):
                    self.status_code = entry.metadata.status_code
                    self.headers = entry.metadata.headers
                    self.url = entry.metadata.url
                    self._data = entry.data
                    self._content = None
                    self._text = None

                @property
                def ok(self) -> bool:
                    return 200 <= self.status_code < 300

                @property
                def is_success(self) -> bool:
                    return self.ok

                def json(self) -> Any:
                    return self._data

                @property
                def text(self) -> str:
                    if self._text is None:
                        if isinstance(self._data, str):
                            self._text = self._data
                        elif self._data is None:
                            self._text = ""
                        else:
                            self._text = json.dumps(self._data)
                    return self._text

                @property
                def content(self) -> bytes:
                    if self._content is None:
                        self._content = self.text.encode("utf-8")
                    return self._content

            return CachedResponse(entry)

    async def invalidate(self, key: str) -> bool:
        """Invalidate a single cache key."""
        return await self._storage.delete(key)

    async def invalidate_pattern(self, pattern: str) -> int:
        """Invalidate keys matching a glob pattern."""
        return await self._storage.delete_pattern(pattern)

    async def clear(self) -> None:
        """Clear all cache entries."""
        await self._storage.clear()

    async def keys(self, pattern: str | None = None) -> list[str]:
        """Get all cache keys."""
        return await self._storage.keys(pattern)

    async def stats(self) -> CacheStats:
        """Get cache statistics."""
        return await self._storage.stats()

    @property
    def stale_while_revalidate_enabled(self) -> bool:
        """Check if stale-while-revalidate is enabled."""
        return self._stale_while_revalidate

    async def close(self) -> None:
        """Close the cache manager."""
        await self._storage.close()

File: platform/packages_mjs/fetch_undici/src/cache/core/cache-manager.ts
/**
 * Cache Manager
 *
 * Core cache logic for managing cached responses.
 */

import { Response, type ResponseInit } from '../../models/response.js'
import type { Headers } from '../../models/headers.js'
import { logger, type Logger } from '../../logger.js'
import { MemoryStorage } from '../storage/memory.js'
import type {
  CacheConfig,
  CacheStorage,
  CacheEntry,
  CacheKeyStrategy,
  RequestCacheOptions
} from '../types.js'
import { defaultKeyStrategy } from './key-strategy.js'

const log = logger.create('fetch-undici', import.meta.url)

/**
 * Cache Manager - Core caching logic
 *
 * Handles cache key generation, storage operations, and response serialization.
 *
 * @example
 * ```typescript
 * const manager = new CacheManager({ ttl: 60000 })
 *
 * const key = manager.generateKey('GET', 'https://api.example.com/users')
 * await manager.set(key, response)
 *
 * const cached = await manager.get(key)
 * const response = manager.createResponseFromCache(cached)
 * ```
 */
export class CacheManager {
  private readonly _storage: CacheStorage
  private readonly _ttl: number
  private readonly _methods: string[]
  private readonly _keyStrategy: CacheKeyStrategy
  private readonly _staleWhileRevalidate: boolean
  private readonly _staleGracePeriod: number
  private readonly _log: Logger

  /** Track in-flight requests for deduplication */
  private readonly _pending = new Map<string, Promise<Response>>()

  constructor(config?: CacheConfig) {
    this._storage =
      config?.storage ??
      new MemoryStorage({
        maxEntries: config?.maxEntries ?? 1000
      })

    this._ttl = config?.ttl ?? 60000
    this._methods = (config?.methods ?? ['GET', 'HEAD']).map((m) => m.toUpperCase())
    this._keyStrategy = config?.keyStrategy ?? defaultKeyStrategy
    this._staleWhileRevalidate = config?.staleWhileRevalidate ?? false
    this._staleGracePeriod = config?.staleGracePeriod ?? 30000
    this._log = config?.logger ?? log

    this._log.debug('CacheManager created', {
      ttl: this._ttl,
      methods: this._methods
    })
  }

  /**
   * Generate a cache key for a request
   */
  generateKey(
    method: string,
    url: string,
    headers?: Record<string, string> | Headers,
    body?: unknown,
    params?: Record<string, unknown>
  ): string {
    const headersObj =
      headers instanceof Map
        ? Object.fromEntries(headers)
        : (headers as Record<string, string> | undefined)

    return this._keyStrategy(method, url, headersObj, body, params)
  }

  /**
   * Check if a request should be cached
   */
  shouldCache(method: string, options?: RequestCacheOptions): boolean {
    if (options?.noCache) return false
    return this._methods.includes(method.toUpperCase())
  }

  /**
   * Get an entry from cache
   */
  async get(key: string): Promise<CacheEntry | undefined> {
    const entry = await this._storage.get(key)

    if (entry) {
      this._log.trace('Cache hit', { key })
    } else {
      this._log.trace('Cache miss', { key })
    }

    return entry
  }

  /**
   * Get with stale-while-revalidate support
   */
  async getStale(key: string): Promise<{
    entry: CacheEntry | undefined
    isStale: boolean
  }> {
    const entry = await this._storage.get(key)

    if (!entry) {
      return { entry: undefined, isStale: false }
    }

    const now = Date.now()
    const isStale = now > entry.expiresAt
    const withinGrace = now < entry.expiresAt + this._staleGracePeriod

    if (isStale && !withinGrace) {
      return { entry: undefined, isStale: true }
    }

    return { entry, isStale }
  }

  /**
   * Store a response in cache
   */
  async set(key: string, response: Response, ttl?: number): Promise<void> {
    const effectiveTtl = ttl ?? this._ttl
    const now = Date.now()

    // Read response body (must be done before caching)
    let data: unknown
    const contentType = response.headers.get('content-type') ?? ''

    try {
      if (contentType.includes('application/json')) {
        data = await response.json()
      } else {
        data = await response.text()
      }
    } catch {
      // If body reading fails, store empty
      data = null
    }

    const entry: CacheEntry = {
      key,
      data,
      createdAt: now,
      expiresAt: now + effectiveTtl,
      metadata: {
        statusCode: response.statusCode,
        headers: this._headersToObject(response.headers),
        url: response.url ?? '',
        method: response.request?.method ?? 'GET',
        contentType: contentType || undefined,
        size: typeof data === 'string' ? data.length : undefined
      }
    }

    await this._storage.set(key, entry)
    this._log.debug('Cached response', { key, ttl: effectiveTtl })
  }

  /**
   * Create a Response from a cached entry
   */
  createResponseFromCache(entry: CacheEntry): Response {
    let body: string

    if (typeof entry.data === 'string') {
      body = entry.data
    } else if (entry.data === null || entry.data === undefined) {
      body = ''
    } else {
      body = JSON.stringify(entry.data)
    }

    const init: ResponseInit = {
      statusCode: entry.metadata.statusCode,
      headers: entry.metadata.headers,
      body
    }

    return new Response(init)
  }

  /**
   * Get or fetch with deduplication
   *
   * Prevents multiple identical requests from hitting the origin.
   */
  async getOrFetch(
    key: string,
    fetchFn: () => Promise<Response>,
    options?: { ttl?: number }
  ): Promise<Response> {
    // Check cache first
    const cached = await this.get(key)
    if (cached) {
      return this.createResponseFromCache(cached)
    }

    // Check if request is already in flight
    const pending = this._pending.get(key)
    if (pending) {
      this._log.trace('Deduplicating request', { key })
      return pending
    }

    // Create new request
    const promise = fetchFn()
    this._pending.set(key, promise)

    try {
      const response = await promise

      // Cache successful responses
      if (response.ok) {
        await this.set(key, response, options?.ttl)
      }

      return response
    } finally {
      this._pending.delete(key)
    }
  }

  /**
   * Invalidate a single cache key
   */
  async invalidate(key: string): Promise<boolean> {
    const deleted = await this._storage.delete(key)
    this._log.debug('Cache invalidated', { key, deleted })
    return deleted
  }

  /**
   * Invalidate keys matching a pattern
   */
  async invalidatePattern(pattern: string | RegExp): Promise<number> {
    const count = await this._storage.deletePattern(pattern)
    this._log.debug('Cache pattern invalidated', { pattern: String(pattern), count })
    return count
  }

  /**
   * Clear all cache entries
   */
  async clear(): Promise<void> {
    await this._storage.clear()
    this._log.info('Cache cleared')
  }

  /**
   * Get all cache keys
   */
  async keys(pattern?: string | RegExp): Promise<string[]> {
    return this._storage.keys(pattern)
  }

  /**
   * Get cache statistics
   */
  async stats() {
    return this._storage.stats()
  }

  /**
   * Check if stale-while-revalidate is enabled
   */
  get staleWhileRevalidateEnabled(): boolean {
    return this._staleWhileRevalidate
  }

  /**
   * Close the cache manager
   */
  async close(): Promise<void> {
    await this._storage.close()
    this._log.debug('CacheManager closed')
  }

  /**
   * Convert Headers to plain object
   */
  private _headersToObject(headers: Headers): Record<string, string> {
    const obj: Record<string, string> = {}
    for (const [key, value] of headers) {
      obj[key] = value
    }
    return obj
  }
}
